packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -15,101 +16,163 @@ import "@opengsn/contracts/src/utils/GsnUtils.sol";
+        tokensBlockNumber = blockNumber;
         // allow uniswap to transfer from paymaster balance
-        token.approve(address(uniswap), type(uint256).max);
+        for (uint256 i = 0; i < tokens.length; i++) {
+            TokenSwapData memory data;
+            IERC20 token = tokens[i];
+            token.approve(address(uniswap), type(uint256).max);
+            data.priceDivisor = 10 ** uint256(_priceFeeds[i].decimals() + IERC20Metadata(address(token)).decimals());
+            data.priceFeed = _priceFeeds[i];
+            data.permitMethodSelector = bytes4(keccak256(bytes(_permitMethodSignatures[i])));
+            data.uniswapPoolFee = _poolFees[i];
+            require(_slippages[i] <= 1000, "slippage above 100%");
+            data.slippage = _slippages[i];
+            data.validFromBlockNumber = blockNumber;
+            tokensSwapData[token] = data;
+        }
     }
 
-    /**
-     * set gas used by postRelayedCall, for proper gas calculation.
-     * You can use TokenGasCalculator to calculate these values (they depend on actual code of postRelayedCall,
-     * but also the gas usage of the token and of Uniswap)
-     */
-    function setPostGasUsage(uint256 _gasUsedByPost) public onlyOwner {
-        gasUsedByPost = _gasUsedByPost;
+    function getTokens() public view returns (IERC20[] memory){
+        return tokens;
     }
 
-    function _calculatePreCharge(
-        GsnTypes.RelayRequest calldata relayRequest,
-        uint256 maxPossibleGas)
-    internal
+    function getTokenSwapData(IERC20 token) public view returns (TokenSwapData memory) {
+        return tokensSwapData[token];
+    }
+
+    function refillHubDeposit(uint256 amount) public payable onlyOwner {
+        _refillHubDeposit(amount);
+    }
+
+    function withdrawTokens(IERC20[] calldata _tokens, address target, uint256[] calldata amounts) public onlyOwner {
+        for (uint256 i = 0; i < _tokens.length; i++) {
+            _tokens[i].safeTransfer(target, amounts[i]);
+        }
+    }
+
+    function _calculateCharge(
+        GsnTypes.RelayData calldata relayData,
+        uint256 gasUsed,
+        uint256 priceQuote,
+        uint256 priceDivisor
+    ) internal
     view
-    returns (uint256 tokenPreCharge) {
-        uint256 ethMaxCharge =
-            relayHub.calculateCharge(maxPossibleGas, relayRequest.relayData) +
-            relayRequest.request.value;
-        uint256 price = uint256(priceFeed.latestAnswer());
-        uint256 decimals = uint256(priceFeed.decimals());
-        tokenPreCharge = ethMaxCharge * price / (10 ** decimals);
+    returns (uint256 tokenCharge, uint256 ethCharge) {
+        ethCharge = relayHub.calculateCharge(gasUsed, relayData);
+        tokenCharge = addPaymasterFee(_weiToToken(ethCharge, priceDivisor, priceQuote));
     }
 
+    function _tokenToWei(uint256 amount, uint256 divisor, uint256 quote) internal pure returns(uint256) {
+        return 1e18 * amount * quote / divisor;
```

Comment:
suggestion: "fold" the divisor into the quote: 
- quote should a "fixed point" value of actualQuote *1e18/divisor
- tokenToWei:  return amount * quote/1e18
- weiToToken: amount*1e18/quote